ledger/internal/apptxn_test.go

```diff
@@ -1500,3 +1506,936 @@ func TestGtxnEffects(t *testing.T) {
+	txn(t, l, eval, &app0)
+	vb := endBlock(t, l, eval)
+	index0 := vb.Block().Payset[0].ApplicationID
+
+	fund0 := txntest.Txn{
+		Type:     "pay",
+		Sender:   addrs[0],
+		Receiver: index0.Address(),
+		Amount:   1_000_000,
+	}
+
+	app1 := txntest.Txn{
+		Type:   "appl",
+		Sender: addrs[0],
+		ApprovalProgram: main(`
+  int 3
+  int 3
+  ==
+  assert
+`),
+	}
+	eval = nextBlock(t, l, true, nil)
+	txns(t, l, eval, &app1, &fund0)
+	vb = endBlock(t, l, eval)
+	index1 := vb.Block().Payset[0].ApplicationID
+
+	callTxGroup := make([]txntest.Txn, 16)
+	for i := 0; i < 16; i++ {
+		callTxGroup[i] = txntest.Txn{
+			Type:          "appl",
+			Sender:        addrs[0],
+			ApplicationID: index0,
+			ForeignApps:   []basics.AppIndex{index1},
+		}
+	}
+	eval = nextBlock(t, l, true, nil)
+	err := txgroup(t, l, eval, &callTxGroup[0], &callTxGroup[1], &callTxGroup[2], &callTxGroup[3], &callTxGroup[4], &callTxGroup[5], &callTxGroup[6], &callTxGroup[7], &callTxGroup[8], &callTxGroup[9], &callTxGroup[10], &callTxGroup[11], &callTxGroup[12], &callTxGroup[13], &callTxGroup[14], &callTxGroup[15])
+	require.Error(t, err)
+
+	endBlock(t, l, eval)
+}
+
+func TestMaxInnerTxForSingleAppCall(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	genBalances, addrs, _ := ledgertesting.NewTestGenesis()
+	l := newTestLedger(t, genBalances)
+	defer l.Close()
+
+	program := `
+int 1
+loop:
+itxn_begin
+  int appl
+  itxn_field TypeEnum
+  txn Applications 1
+  itxn_field ApplicationID
+itxn_submit
+int 1
++
+dup
+int 256
+<=
+bnz loop
+int 257
+==
+assert
+`
+
+	// 256 inner txns
+	app0 := txntest.Txn{
+		Type:            "appl",
+		Sender:          addrs[0],
+		ApprovalProgram: main(program),
+	}
+	eval := nextBlock(t, l, true, nil)
+	txn(t, l, eval, &app0)
+	vb := endBlock(t, l, eval)
+	index0 := vb.Block().Payset[0].ApplicationID
+
+	fund0 := txntest.Txn{
+		Type:     "pay",
+		Sender:   addrs[0],
+		Receiver: index0.Address(),
+		Amount:   1_000_000,
+	}
+
+	app1 := txntest.Txn{
+		Type:   "appl",
+		Sender: addrs[0],
+		ApprovalProgram: main(`
+  int 3
+  int 3
+  ==
+  assert
+`),
+	}
+	eval = nextBlock(t, l, true, nil)
+	txns(t, l, eval, &app1, &fund0)
+	vb = endBlock(t, l, eval)
+	index1 := vb.Block().Payset[0].ApplicationID
+
+	callTxGroup := make([]txntest.Txn, 16)
+	callTxGroup[0] = txntest.Txn{
+		Type:          "appl",
+		Sender:        addrs[0],
+		ApplicationID: index0,
+		ForeignApps:   []basics.AppIndex{index1},
+	}
+	for i := 1; i < 16; i++ {
+		callTxGroup[i] = txntest.Txn{
+			Type:          "appl",
+			Sender:        addrs[0],
+			ApplicationID: index1,
+			ForeignApps:   []basics.AppIndex{},
+		}
+	}
+	eval = nextBlock(t, l, true, nil)
+	err := txgroup(t, l, eval, &callTxGroup[0], &callTxGroup[1], &callTxGroup[2], &callTxGroup[3], &callTxGroup[4], &callTxGroup[5], &callTxGroup[6], &callTxGroup[7], &callTxGroup[8], &callTxGroup[9], &callTxGroup[10], &callTxGroup[11], &callTxGroup[12], &callTxGroup[13], &callTxGroup[14], &callTxGroup[15])
```

Comment:
Since we decided to allow 256 calls even from a singleton app call (not a 16 app group) we should adjust this test.  I'll let you know when I have that merged in.